### PR TITLE
Communicate with Ryuk over the Docker connection

### PR DESCRIPTION
**Pull request by bsideup**
Related to #3136

### CODE REVIEW

File: core/src/main/java/org/testcontainers/utility/ResourceReaper.java [hunk 1]
```diff
@@ -70,118 +55,129 @@ private ResourceReaper() {
...
-
-        synchronized (DEATH_NOTE) {
-            DEATH_NOTE.add(
-                    DockerClientFactory.DEFAULT_LABELS.entrySet().stream()
-                            .<Map.Entry<String, String>>map(it -> new SimpleEntry<>("label", it.getKey() + "=" + it.getValue()))
-                            .collect(Collectors.toList())
-            );
+        if (!attachment.awaitStarted(TestcontainersConfiguration.getInstance().getRyukTimeout(), TimeUnit.SECONDS)) {
+            log.error("Timed out waiting for Ryuk container's attach. Container id: '{}'", ryukContainerId);
+            throw new IllegalStateException("Failed to attack too Ryuk");
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/utility/ResourceReaper.java:146]: Tiny typos!
```suggestion
            throw new IllegalStateException("Failed to attach to Ryuk");
```

Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/ResourceReaper.java:146]: omg :D thanks!

Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/ResourceReaper.java]: what a combo of Mac's keyboard and my late evening brain 😂

File: core/src/main/java/org/testcontainers/utility/ResourceReaper.java [hunk 2]
```diff
@@ -70,118 +55,129 @@ private ResourceReaper() {
...
+        log.debug("Starting Ryuk container with id: '{}'", ryukContainerId);
+        client.startContainerCmd(ryukContainerId).exec();
+
         try {
-            // We need to wait before we can start any containers to make sure that we delete them
-            if (!ryukScheduledLatch.await(TestcontainersConfiguration.getInstance().getRyukTimeout(), TimeUnit.SECONDS)) {
+            if (!started.await(TestcontainersConfiguration.getInstance().getRyukTimeout(), TimeUnit.SECONDS)) {
                 log.error("Timed out waiting for Ryuk container to start. Ryuk's logs:\n{}", ryukLog);
-                throw new IllegalStateException(String.format("Could not connect to Ryuk at %s:%s", hostIpAddress, ryukPort));
+                throw new IllegalStateException("Could not connect to Ryuk");
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/utility/ResourceReaper.java:155]: Should we take the opportunity to make this clearer?
e.g.
```suggestion
                throw new IllegalStateException("Timed out waiting for the Ryuk cleanup container to start");
```

Comment (bsideup) [reply core/src/main/java/org/testcontainers/utility/ResourceReaper.java:155]: maybe "watchdog container" instead of "cleanup container"?

**Comment by asafm** [comment]
@bsideup Sorry for late response - what's the trick again to try this commit as dependency?

**Comment by asafm** [comment]
@bsideup Ok, finally got some time to run it. If fails.
```
20/12/18 22:03:49.411 [testcontainers-ryuk] WARN o.t.u.ResourceReaper: Ryuk connection failure
java.lang.RuntimeException: java.net.SocketException: Socket closed
	at com.github.dockerjava.api.async.ResultCallbackTemplate.throwFirstError(ResultCallbackTemplate.java:158)
	at com.github.dockerjava.api.async.ResultCallbackTemplate.awaitCompletion(ResultCallbackTemplate.java:93)
	at org.testcontainers.utility.ResourceReaper.lambda$null$1(ResourceReaper.java:105)
	at org.testcontainers.utility.ResourceReaper$$Lambda$542/1971736492.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.SocketException: Socket closed
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:170)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at org.testcontainers.shaded.okio.Okio$2.read(Okio.java:140)
	at org.testcontainers.shaded.okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
	at org.testcontainers.shaded.okio.RealBufferedSource.read(RealBufferedSource.java:51)
	at org.testcontainers.shaded.okhttp3.internal.http1.Http1ExchangeCodec$AbstractSource.read(Http1ExchangeCodec.java:389)
	at org.testcontainers.shaded.okhttp3.internal.http1.Http1ExchangeCodec$UnknownLengthSource.read(Http1ExchangeCodec.java:529)
	at org.testcontainers.shaded.okhttp3.internal.connection.Exchange$ResponseBodySource.read(Exchange.java:286)
	at org.testcontainers.shaded.okio.RealBufferedSource$1.read(RealBufferedSource.java:436)
	at org.testcontainers.shaded.com.github.dockerjava.core.FramedInputStreamConsumer.accept(FramedInputStreamConsumer.java:30)
	at org.testcontainers.shaded.com.github.dockerjava.core.FramedInputStreamConsumer.accept(FramedInputStreamConsumer.java:12)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.lambda$executeAndStream$1(DefaultInvocationBuilder.java:275)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder$$Lambda$380/664856708.run(Unknown Source)
	... 1 common frames omitted

```
on ```java.lang.RuntimeException: java.lang.IllegalStateException: Timed out waiting for the default Ryuk filters to be registered```

I'll try to work on this some more to figure out why the socket was closed. I think it's the DockerClient socket.

**Comment by bsideup** [comment]
Going to rework

**Pull request closed without merging**